text/0000-io-net-2.1.md

```diff
@@ -0,0 +1,527 @@
+    /// syscall is set to 0.
+    pub fn send_to<A: ToSocketAddrs>(&self, data: &[u8], addr: A)
+                                     -> io::Result<usize>;
+
+    /// Same as `UdpSocket::recv_from`, the `flags` parameter to the underlying
+    /// syscall is set to 0.
+    pub fn recv_from(&self, buf: &mut [u8]) -> io::Result<(usize, SocketAddr)>;
+}
+
+// Standard I/O operations
+impl Read for Socket { ... }
+impl Read for &Socket { ... }
+impl Write for Socket { ... }
+impl Write for &Socket { ... }
+
+// Conversion traits/types for `Socket`
+impl AsRaw{Fd,Socket} for Socket { ... }
+impl From<Socket> for TcpListener { ... }
+impl From<Socket> for TcpStream { ... }
+impl From<Socket> for UdpSocket { ... }
+impl<'a> Into<RawSocketAddr> for &'a SocketAddr { ... }
+
+// Common constructors, conversions, and inspectors for the various arguments to
+// `Socket::new`. Platform-specific constructors can be provided through
+// platform specific extension traits
+impl Domain {
+    pub fn ipv4() -> Domain;
+    pub fn ipv6() -> Domain;
+    pub fn value(&self) -> c_int;
+}
+impl From<c_int> for Domain { ... }
+
+impl Type {
+    pub fn stream() -> Type;
+    pub fn datagram() -> Type;
+    pub fn raw() -> Type;
+    pub fn value(&self) -> c_int;
+}
+impl From<c_int> for Type { ... }
+
+impl From<c_int> for Protocol { ... }
+```
+
+The upside of this approach is that it is maximally flexible in terms of usage.
+All interactions with C APIs can be directly translated to Rust itself with ease
+and there is a clear layer of abstraction for what's going on here. The
+downsides of this approach, however, are:
+
+* It's unclear why the `TcpStream` and other refined primitives still exist.
+  This API is a union of the `TcpStream`, `TcpListener`, and `UdpSocket` APIs
+  and poses a new question of "should I use a Socket or TcpStream?".
+* The scope of this API is somewhat unclear, for example is `Socket` intended to
+  encompass Unix sockets as well? It attempts to via the `Into<RawSocketAddr>`
+  trait bound, but it's arguably not super ergonomic.
+* Many mis-usages of the API which would statically be prevented (such as
+  calling `accept` on a UDP socket) are allowed with a single `Socket` API.
+  Other examples of this are setting TCP options for a datagram socket.
+
+Overall the builder-based approach seemed to give the same level of flexibility
+while maintaining a clearly defined purpose of solely operating as builders for
+new sockets.
+
+### A note on typestate
+
+Note that there are many alternatives within this alternative itself (of having
+a unifying notion of a "socket"), such as using typestate to solve the mis-usage
+bullet point above. Sockets are often a classical use case of a state-transition
+diagram, so this can be a natural conclusion to come to.
+
+There are, however, not that many usages of typestate throughout the rest of the
+standard library today, and there are a number of ergonomic and API concerns
```

Comment:
Without designing the whole thing, but to give a peek, it could look something like this:

```rust
struct Socket<Type> {
    // ...
}

enum Tcp {}
enum Udp {}

impl<T> Socket<T> {
    // any socket has these methods
    pub fn try_close() {}
    pub fn shutdown() {}
}
impl Socket<Tcp> {
    pub fn accept(&mut self) -> io::Result<()> {
        // ...
    }
}
```

Usage then could be:

```rust
let mut sock = Socket::<Udp>::connect(addr);
let other = sock.try_clone().unwrap();

sock.accept(); // compile error, Socket<Udp> does not have a method `accept`
```